tests/dragonfly/replication_test.py

```diff
@@ -2411,3 +2411,66 @@ async def test_replicate_old_master(
     assert await c_replica.execute_command("get", "k1") == "v1"
 
     await disconnect_clients(c_master, c_replica)
+
+
+@pytest.mark.asyncio
+async def test_empty_hash_map_replicate_old_master(df_factory):
+    cpu = platform.processor()
+    if cpu != "x86_64":
+        pytest.skip(f"Supported only on x64, running on {cpu}")
+
+    dfly_version = "v1.21.2"
+    released_dfly_path = download_dragonfly_release(dfly_version)
+
+    master = df_factory.create(path=released_dfly_path)
+    master.clear_max_chunk_flag()
+    master.start()
+
+    replica = df_factory.create(path=released_dfly_path)
+    replica.clear_max_chunk_flag()
+    replica.start()
+
+    c_master = master.client()
+    # Create an empty hashmap
+    await c_master.execute_command("HSET foo a_field a_value")
+    await c_master.execute_command("HSETEX foo 2 b_field b_value")
+    await c_master.execute_command("HDEL foo a_field")
+
+    @assert_eventually
+    async def check_if_empty():
+        assert await c_master.execute_command("HGETALL foo") == []
+
+    await check_if_empty()
+    assert await c_master.execute_command(f"EXISTS foo") == 1
+
+    c_replica = replica.client()
+    # Round one
+    assert await c_replica.execute_command(f"REPLICAOF localhost {master.port}") == "OK"
+    async with async_timeout.timeout(10):
+        await wait_for_replicas_state(c_replica, state="stable_sync", node_role="replica")
+
+    assert await c_replica.execute_command(f"EXISTS foo") == 1
+    # Replica is now master
+    await c_replica.execute_command("REPLTAKEOVER 1")
+    master.start()
+    c_master = master.client()
+    # Master is now replica
+    await c_master.execute_command(f"REPLICAOF localhost {replica.port}")
+    async with async_timeout.timeout(10):
+        await wait_for_replicas_state(c_master, state="stable_sync", node_role="replica")
+    # Promote master back to master
+    await c_master.execute_command("REPLTAKEOVER 1")
+
+    # New replica with latest version that includes the bug fix -- we should be able to load
+    replica = df_factory.create()
+    replica.start()
+    c_replica = replica.client()
+    await c_replica.execute_command(f"REPLICAOF localhost {master.port}")
+    async with async_timeout.timeout(10):
+        await wait_for_replicas_state(c_replica)
+    assert await c_replica.execute_command(f"EXISTS foo") == 0
+    assert await c_replica.execute_command("REPLTAKEOVER 1") == "OK"
+    role, node = await c_replica.execute_command("role")
+    assert role == "master"
+
+    await close_clients(c_master, c_replica)
```

Comment:
you could've just spawned two more replicas (current version) and done to avoid this manual code

```py
instances = [old1, old2, old3, new1]
for a, b in zip(instances, instances[1:]):
  # replicate a to b
```